File: src/java/BinarySearch.java
/**
 * Copyright (C) 2016 Canux CHENG.
 * All right reserved.
 * Name: BinarySearch.java
 */

//package jva;

/**
 * Binary search
 * @author Canux CHENG canuxcheng@gmail.com
 * @version V1.0.0.0
 * @since Sat 17 Sep 2016 07:12:01 AM EDT
 */
public class BinarySearch
{
    public static void main(String[] args)
    {
        int[] array = {32, 19, -2, 48, 90};
        System.out.println(binarySearch(array, 0));
        System.out.println(binarySearch(array, -2));
    }

    /**
     * binary search for one value.
     * @param array.
     * @param key
     * @return mid
     */
    public static int binarySearch(int[] array, int key)
    {
        int min = 0;
        int max = array.length - 1;

        while (min <= max)
        {
            int mid = (min + max)>>1;

            if (key > array[mid])
            {
                min = mid + 1;
            }
            else if (key < array[mid])
            {
                max = mid - 1;
            }
            else
                return mid;
        }
        return -1;
    }
}

File: algorithm/binary_search.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
binary_search.py for search algorithm binary.

Copyright (C) 2017 Canux CHENG.
All rights reserved.

LICENSE GNU General Public License v3.0.

:author: Canux CHENG canuxcheng@gmail.com
:version: V1.0.0.0
:since: Fri 20 Jan 2017 01:10:07 AM EST

DESCRIPTION:
"""


def binary_search(seq, key):
    """Main function use binary search algorithm to find the key word in a sequence.

    :param seq: sequence can be string, tupple, list.
    :type seq: sequence.
    :param key: an element of a sequence.
    :type key: char or int/float/complex.
    :return mid_index: the index in seq for key.
    :rtype mid_index: int.
    """
    min_index = 0
    max_index = len(seq) - 1

    while min_index <= max_index:
        # mid_index = min_index + (max_index - min_index) // 2
        mid_index = (min_index + max_index) >> 1
        if key > seq[mid_index]:
            min_index = mid_index + 1
        elif key < seq[mid_index]:
            max_index = mid_index - 1
        else:
            return mid_index
    return False


if __name__ == "__main__":
    seq = [-9, 6, 15, 159, 2489, 98756]
    key = 159
    # seq = "abcdxyz"
    # key = 'd'
    print("index {0} in {1} is the same with {2}.".format(binary_search(seq, key), seq, key))
